client/core/trade.go

```diff
@@ -1688,28 +1700,75 @@ func (t *trackedTrade) processRedemption(msgID uint64, redemption *msgjson.Redem
 	return nil
 }
 
-// hasUnswappedMatches will be true if the order has any remaining matches for
-// which our swap has not been broadcast. hasUnswappedMatches should be called
-// with the matchMtx locked.
-func (t *trackedTrade) hasUnswappedMatches() bool {
+// Coins will be returned if
+// - the trade status is not OrderStatusEpoch or OrderStatusBooked, that is to
+//   say, there won't be future matches for this order.
+// - there are no matches in the trade that MAY later require sending swaps,
+//   that is to say, all matches have been either swapped or revoked.
+//
+// This method modifies match fields and MUST be called with the trackedTrade
+// mutex lock held for writes.
+func (t *trackedTrade) maybeReturnCoins() bool {
+	// Status of the order itself.
+	if t.metaData.Status < order.OrderStatusExecuted {
+		// Booked and epoch orders may get matched any moment from
+		// now, keep the coins locked.
+		log.Tracef("not unlocking coins for order with status %s", t.metaData.Status)
+		return false
+	}
+
+	// Status of all matches for the order. If a match exists for
+	// which a swap MAY be sent later, keep the coins locked.
 	for _, match := range t.matches {
-		dbMatch := match.Match
-		if (dbMatch.Side == order.Taker && dbMatch.Status < order.TakerSwapCast) ||
-			(dbMatch.Side == order.Maker && dbMatch.Status < order.MakerSwapCast) {
+		if match.MetaData.Proof.IsRevoked {
+			// Won't be sending swap for this match regardless of
+			// the match's status.
+			continue
+		}
 
-			return true
+		status, side := match.Match.Status, match.Match.Side
+		if side == order.Maker && status < order.MakerSwapCast ||
+			side == order.Taker && status < order.TakerSwapCast {
+			// Match is active (not revoked, not refunded) and client
+			// is yet to execute swap. Keep coins locked.
+			log.Tracef("not unlocking coins for order with match side %s, status", side, status)
+			return false
 		}
 	}
-	return false
+
+	// Safe to unlock coins now.
+	t.returnCoins()
+	return true
 }
 
-// coinList makes a slice of the coin map values.
-func (t *trackedTrade) coinList() []asset.Coin {
-	coins := make([]asset.Coin, 0, len(t.coins))
-	for _, coin := range t.coins {
-		coins = append(coins, coin)
+// returnCoins unlocks this trade's funding coins (if unspent) or the change
+// coin if a previous swap created a change coin that is locked.
+// Coins are auto-unlocked once spent in a swap tx, including intermediate
+// change coins, such that only the last change coin (if locked), will need
+// to be unlocked.
+//
+// This method modifies match fields and MUST be called with the trackedTrade
+// mutex lock held for writes.
+func (t *trackedTrade) returnCoins() {
```

Comment:
Will need a way to communicate to `Core` to refresh the fromWallet balance after unlocking coins. Probably a better candidate for #672. Will add a comment there.

Reply:
The caller of `returnCoins` be doing this, as in `tick`?  We're relying on tick to do this now though it seems

The DB failure with `UpdateOrder` is hard to handle because it basically means the program is totally hosed.  I don't have a good resolution in mind.

Reply:
`returnCoins` may be called from a number of places aside `tick`. I'm not sure how I feel about having all callers of `returnCoins` also refresh wallet balance but will think this through more when working on #672.